drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,807 @@
+	LOG_DBG("buffer=0x%x", (uint32_t)buffer);
+	if (buffer == NULL) {
+		goto error;
+	}
+
+	if (net_pkt_read(pkt, buffer, total_len)) {
+		goto error;
+	}
+
+	/* Prepare transmit descriptors to give to DMA */
+	m_numaker_gmacdev_trigger_tx(gmacdev, total_len);
+
+	k_mutex_unlock(&data->tx_frame_buf_mutex);
+
+	return 0;
+
+error:
+	LOG_ERR("Writing pkt to TX descriptor failed");
+	k_mutex_unlock(&data->tx_frame_buf_mutex);
+	return -EIO;
+}
+
+static void numaker_eth_if_init(struct net_if *iface)
+{
+	const struct device *dev = net_if_get_device(iface);
+	struct eth_numaker_data *data = dev->data;
+
+	synopGMACdevice *gmacdev = data->gmacdev;
+
+	LOG_DBG("eth_if_init");
+
+	/* Read mac address */
+	m_numaker_read_mac_addr(data->mac_addr);
+
+	net_if_set_link_addr(iface, data->mac_addr, sizeof(data->mac_addr), NET_LINK_ETHERNET);
+	data->iface = iface;
+	ethernet_init(iface);
+
+	/* Enable GMAC device INT & TX/RX */
+	m_numaker_gmacdev_enable(gmacdev);
+}
+
+static int numaker_eth_set_config(const struct device *dev, enum ethernet_config_type type,
+				  const struct ethernet_config *config)
+{
+	struct eth_numaker_data *data = dev->data;
+
+	switch (type) {
+	case ETHERNET_CONFIG_TYPE_MAC_ADDRESS:
+		memcpy(data->mac_addr, config->mac_address.addr, sizeof(data->mac_addr));
+		synopGMAC_set_mac_address(NUMAKER_GMAC_INTF, data->mac_addr);
+		net_if_set_link_addr(data->iface, data->mac_addr, sizeof(data->mac_addr),
+				     NET_LINK_ETHERNET);
+		LOG_DBG("%s MAC set to %02x:%02x:%02x:%02x:%02x:%02x", dev->name, data->mac_addr[0],
+			data->mac_addr[1], data->mac_addr[2], data->mac_addr[3], data->mac_addr[4],
+			data->mac_addr[5]);
+		return 0;
+	default:
+		break;
+	}
+
+	return -ENOTSUP;
+}
+
+static enum ethernet_hw_caps numaker_eth_get_cap(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+#if defined(NU_USING_HW_CHECKSUM)
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T | ETHERNET_HW_RX_CHKSUM_OFFLOAD;
+#else
+	return ETHERNET_LINK_10BASE_T | ETHERNET_LINK_100BASE_T;
+#endif
+}
+
+static const struct ethernet_api eth_numaker_driver_api = {
+	.iface_api.init = numaker_eth_if_init,
+	.get_capabilities = numaker_eth_get_cap,
+	.set_config = numaker_eth_set_config,
+	.send = numaker_eth_tx,
+};
+
+/* EMAC IRQ Handler */
+static void eth_numaker_isr(const struct device *dev)
+{
+	struct eth_numaker_data *data = dev->data;
+	synopGMACdevice *gmacdev = data->gmacdev;
+	uint32_t interrupt;
+	uint32_t dma_status_reg;
+	uint32_t mac_status_reg;
+	int status;
+	uint32_t dma_ie = DmaIntEnable;
+
+	uint32_t volatile reg;
+
+	/* Check GMAC interrupt */
+	mac_status_reg = synopGMACReadReg((u32 *)gmacdev->MacBase, GmacInterruptStatus);
+	if (mac_status_reg & GmacTSIntSts) {
+		gmacdev->synopGMACNetStats.ts_int = 1;
+		status = synopGMACReadReg((u32 *)gmacdev->MacBase, GmacTSStatus);
+		if (!(status & (1 << 1))) {
+			LOG_DBG("TS alarm flag not set??");
+		} else {
+			LOG_DBG("TS alarm!!!!!!!!!!!!!!!!");
```

Comment:
nit: drop some `!` save some flash, I appreceate the overly enthusiastic debug messages but that does not add any information :-)

Reply:
Hi @fabiobaltieri ,
Thanks of your advice.
I updated all of above advice at commit be3802c105